### PR TITLE
fix emit for super calls that has preceding prologue directives

### LINKED ISSUE #7877: [Transforms] field initializers are emitted before super call in presence of string directives
**Expected**:

```js
//// [emitSuperCallBeforeEmitPropertyDeclarationAndParameterPropertyDeclaration1ES6.ts]
class A {
    blub = 6;
}


class B extends A {
    blah = 2;
    constructor(public x: number) {
        "use strict";
        'someStringForEgngInject';
        super()
    }
}

//// [emitSuperCallBeforeEmitPropertyDeclarationAndParameterPropertyDeclaration1ES6.js]
class A {
    constructor() {
        this.blub = 6;
    }
}
class B extends A {
    constructor(x) {
        "use strict";
        'someStringForEgngInject';
        super();
        this.x = x;
        this.blah = 2;
    }
}
```

**Actual**:

```js
//// [emitSuperCallBeforeEmitPropertyDeclarationAndParameterPropertyDeclaration1ES6.ts]
class A {
    blub = 6;
}


class B extends A {
    blah = 2;
    constructor(public x: number) {
        "use strict";
        'someStringForEgngInject';
        super()
    }
}

//// [emitSuperCallBeforeEmitPropertyDeclarationAndParameterPropertyDeclaration1ES6.js]
class A {
    constructor() {
        this.blub = 6;
    }
}
class B extends A {
    constructor(x) {
        this.x = x;
        this.blah = 2;
        "use strict";
        'someStringForEgngInject';
        super();
    }
}
```

Failed tests:
* tests\cases\compiler\emitSuperCallBeforeEmitPropertyDeclarationAndParameterPropertyDeclaration1ES6.ts
* tests\cases\compiler\strictModeInConstructor.ts

### PR DESCRIPTION
**Pull request by vladima**
don't assume that super call is the first statement in the list. Fixes #7877.

### CODE REVIEW

File: src/compiler/transformers/ts.ts [hunk 1]
```diff
@@ -841,25 +836,44 @@ namespace ts {
...
 
         /**
-         * Finds the initial super-call for a constructor.
+         * Adds super call and preceding prologue directives into the list of statements.
          *
          * @param ctor The constructor node.
+         * @returns index of the statement that follows super call
          */
-        function findInitialSuperCall(ctor: ConstructorDeclaration): ExpressionStatement {
+        function addPrologueDirectivesAndInitialSuperCall(ctor: ConstructorDeclaration, result: Statement[]): number {
```
Comment (rbuckton) [inline src/compiler/transformers/ts.ts]: There's already an `addPrologueDirectives` function. Can we that instead?

Comment (vladima) [inline src/compiler/transformers/ts.ts]: sure

**Comment by vladima** [comment]
@rbuckton any other comments?

File: src/compiler/transformers/ts.ts [hunk 2]
```diff
@@ -849,25 +844,36 @@ namespace ts {
...
+                if (index === statements.length) {
+                    // list contains nothing but prologue directives (or empty) - exit
+                    return index;
+                }
+
+                const statement = statements[index];
+                if (statement.kind === SyntaxKind.ExpressionStatement) {
                     const expression = (<ExpressionStatement>statement).expression;
                     if (expression.kind === SyntaxKind.CallExpression) {
                         if ((<CallExpression>expression).expression.kind === SyntaxKind.SuperKeyword) {
```
Comment (DanielRosenwasser) [inline src/compiler/transformers/ts.ts]: There is an `isSuperCallExpression` function.

**Comment by DanielRosenwasser** [comment]
:+1: apart from the last comment I left.

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**